### PR TITLE
Migration agent-info data to Wazuh DB implementation

### PR DESCRIPTION
**Pull request by DProvinciani**
|Related issue|
|---|
|[Issue 5367](https://github.com/wazuh/wazuh/issues/5367)|

## Description

This PR includes all the changes developed as part of the epic: **[Migration agent-info data to wazuhdb](https://github.com/wazuh/wazuh/issues/5367)**. These changes are:

- **global.db managed by Wazuh DB**: Now, `Wazuh DB` has the capability of access `global.db` for either query, update or modify it. This is simplified by a set of methods that were already in the code but were modified in order to avoid accessing `global.db` in a regular direct way to instead make use of the `Wazuh DB` socket interface.

**Note**: When upgrading, the old `global.db` database is relocated from the `var/db/` folder to `queue/db/`. In addition, the database permissions and ownership are set in order to make Wazuh DB be able to manage it. Finally, in the first manager initialization after the upgrade, the database is updated to the new schema by adding columns and tables.

- **remoted no longer creates agent-info files**: The agent control messages that were stored in the `agent-info` folder in plain text files are now saved in the `global.db`. This mechanism already existed but was made in the past by `wazuh-modulesd:database`. Now the action is performed by `remoted` and no plain text files are created in the `agent-info` folder.

- **new `sync_status` column in `global.db`**: There is a new column in `global.db` called `sync_status` that is used to reflect the synchronization status of the data in the cluster. This column is mainly used in the workers where each keepalive sets it to `WDB_SYNC_REQ`, and then, when the cluster requires to `Wazuh DB` the agents to be synchronized, it is set to `WDB_SYNCED` following the next state machine.

image

- **`wazuh db` commands for cluster synchronization**: There are two new commands in `Wazuh DB` that are used for the agents' data synchronization in the cluster. The first one is `sync-agent-info-get` and is used in the `worker` nodes. It will return all the agents in `global.db` with the `sync_status` in `WDB_SYNC_REQ`. At the same time, and before returning the data, it will set that value to `WDB_SYNCED`. The second one is `sync-agent-info-set` and is used in the `master` nodes. It will insert all the agents' data returned by the previous one. This allows the cluster to implement the next sequence of tasks.

image

- **`Vulnerability Detector` working with `Wazuh DB`**: There was a query made by `Vulnerability Detector` to `global.db` to get the list of agents connected in order to perform the scans. Now, this query is made trough `Wazuh DB`.

- **`agent_control` working with `Wazuh DB`**: The `agent_control` tool no longer uses the `agent-info` plain text files. Now, it gets the information from `global.db` by executing one of the new `Wazuh DB` commands.

- **`manage_agents` working with `Wazuh DB`**: The `manage_agents` tool no longer uses the `agent-info` plain text files. Now, it gets the information from `global.db` by executing one of the new `Wazuh DB` commands.

With the changes introduced in this pull request, a simplified architecture diagram looks like:

image

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [ ] Windows
  - [ ] MAC OS X
- [x] Source installation
- [x] Package installation
- [x] Source upgrade
- [x] Package upgrade
- [ ] Review logs syntax and correct language
- [ ] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)
  - [x] Dr. Memory
  - [x] AddressSanitizer
- Memory tests for Windows
  - [ ] Scan-build report
  - [ ] Coverity
  - [ ] Dr. Memory
- Memory tests for macOS
  - [ ] Scan-build report
  - [ ] Leaks
  - [ ] AddressSanitizer

- [ ] Retrocompatibility with older Wazuh versions
- [ ] Working on cluster environments
- [ ] Configuration on demand reports new parameters
- [ ] The data flow works as expected (agent-manager-api-app)
- [ ] Added unit tests (for new features)
- [ ] Stress test for affected components

### CODE REVIEW

File: src/addagent/validate.c [hunk 1]
```diff
@@ -182,9 +182,12 @@ int OS_RemoveAgent(const char *u_id) {
     if (sock >= 0) {
         close(sock);
     }
-
     os_free(wdboutput);
 
+    if(wdb_remove_agent(atoi(u_id)) != OS_SUCCESS){
```
Comment (vikman90) [inline src/addagent/validate.c:187]: Add a space after `if`.

File: src/init/wazuh/wazuh.sh [hunk 2]
```diff
@@ -97,13 +97,22 @@ WazuhUpgrade()
...
-    # Remove existing SQLite databases
-
-    rm -f $DIRECTORY/var/db/global.db*
+    # Remove/relocate existing SQLite databases
     rm -f $DIRECTORY/var/db/.profile.db*
     rm -f $DIRECTORY/var/db/.template.db*
     rm -f $DIRECTORY/var/db/agents/*
 
+    if [ -f "$DIRECTORY/var/db/global.db" ]; then
+        cp $DIRECTORY/var/db/global.db* $DIRECTORY/queue/db/
```
Comment (vikman90) [inline src/init/wazuh/wazuh.sh:106]: We should move `global.db` only. Discard SHM and WAL files if any.

File: src/remoted/manager.c [hunk 3]
```diff
@@ -88,17 +90,35 @@ void cleaner(void* data){
...
 void save_controlmsg(const keyentry * key, char *r_msg, size_t msg_length)
 {
-    char msg_ack[OS_FLSIZE + 1];
-    char *end;
-    char *uname = "";
-    pending_data_t *data;
-    FILE * fp;
-    mode_t oldmask;
+    char msg_ack[OS_FLSIZE + 1] = "";
+    char *msg = NULL;
```
Comment (vikman90) [inline src/remoted/manager.c:94]: Should we group all these variables into a struct?

File: src/remoted/manager.c [hunk 4]
```diff
@@ -1247,9 +1279,8 @@ void *update_shared_files(__attribute__((unused)) void *none) {
 
 void free_pending_data(pending_data_t *data) {
     if (!data) return;
-    if (data->message) free(data->message);
-    if (data->keep_alive) free(data->keep_alive);
-    free(data);
+    os_free(data->message);
```
Comment (vikman90) [inline src/remoted/manager.c:1282]: Why set `data->message` and `data` to `NULL`, if we're losing them?

File: src/shared/read-agents.c [hunk 5]
```diff
@@ -1102,262 +1094,233 @@ static int _get_time_rkscan(const char *agent_name, const char *agent_ip, agent_
...
-    }
-
-    snprintf(tmp_file, 512, "%s/%s-%s", AGENTINFO_DIR, agent_name, agent_ip);
-
-    /* Set back the IP address */
-    if (agent_ip_pt) {
-        *agent_ip_pt = '/';
+    
+    json_field = cJSON_GetObjectItem(json_agt_info->child, "last_keepalive");
+    if(cJSON_IsNumber(json_field)){
```
Comment (vikman90) [inline src/shared/read-agents.c:1157]: Add a space after `if`.

File: src/shared/read-agents.c [hunk 6]
```diff
@@ -1102,262 +1094,233 @@ static int _get_time_rkscan(const char *agent_name, const char *agent_ip, agent_
...
+        last_keepalive = json_field->valueint;
+        cJSON_Delete(json_agt_info);
+    
+    } else{
```
Comment (vikman90) [inline src/shared/read-agents.c:1161]: Missing space after `else`.

File: src/shared/read-agents.c [hunk 7]
```diff
@@ -1102,262 +1094,233 @@ static int _get_time_rkscan(const char *agent_name, const char *agent_ip, agent_
...
+char **get_agents_by_last_keepalive(int flag, int delta){
+    size_t array_size = 0;
+    char **agents_array = NULL;
+    int *id_array = NULL;
+    int i = 0;
+    cJSON *json_agt_info = NULL;
+    cJSON *json_name = NULL;
+    cJSON *json_ip = NULL;
+
+    switch(flag){
```
Comment (vikman90) [inline src/shared/read-agents.c:1279]: Missing space after `switch`.

File: src/shared/read-agents.c [hunk 8]
```diff
@@ -1102,262 +1094,233 @@ static int _get_time_rkscan(const char *agent_name, const char *agent_ip, agent_
...
+            break;
+        case GA_ACTIVE:
+            id_array = wdb_get_agents_by_keepalive(">", time(0)-delta, FALSE);
+            break;
+        default:
+            mdebug1("Invalid flag '%d' trying to get agents.", flag);
+            return NULL;
+    }
+
+    if(!id_array){
```
Comment (vikman90) [inline src/shared/read-agents.c:1291]: Missing space after `if`.

File: src/shared/remoted_op.c [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+            os_strdup(ARCHS[i], os_arch);
+            break;
+        }
+    }
+
+    if (!ARCHS[i]) {
+        os_strdup("", os_arch);
+    }
+
+    mdebug2("Detected architecture from %s: %s", os_header, os_arch);
```
Comment (vikman90) [inline src/shared/remoted_op.c:40]: This message is prone to flood the log.

File: src/shared/remoted_op.c [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+ * @param[in] msg The agent update message string to be parsed.
+ * @param[Out] os_name The OS name.
+ * @param[Out] os_major The OS major version.
+ * @param[Out] os_minor The OS minor version.
+ * @param[Out] os_build The OS build number.
+ * @param[Out] os_version The OS full version.
+ * @param[Out] os_codename The OS codename.
+ * @param[Out] os_platform The OS platform.
+ * @param[Out] os_arch The OS architecture.
+ */
```
Comment (vikman90) [inline src/shared/remoted_op.c:57]: We should include an `@pre` sentence about every output pointer must be valid.

File: src/shared/remoted_op.c [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+    os_strdup(msg, msg_tmp);
+
+    for (line = strtok_r(msg_tmp, sdelim, &savedptr); line; line = strtok_r(NULL, sdelim, &savedptr)) {
+        switch (*line) {
+        case '#':  // System label
+        case '!':  // Hidden label
+        case '\"': // Regular label
+            // The _agent_ip will not be appended to the labels string.
+            // Instead it will be returned in the agent_ip parameter.
+            if(!strncmp(line, agent_ip_label, strlen(agent_ip_label))) {
```
Comment (vikman90) [inline src/shared/remoted_op.c:225]: Missing space after `if`.

**Review by vikman90 — CHANGES_REQUESTED** [review]

File: src/addagent/validate.c (hunk 1 above)
Comment (DProvinciani) [reply src/addagent/validate.c:187]: FIXED

File: src/init/wazuh/wazuh.sh (hunk 2 above)
Comment (DProvinciani) [reply src/init/wazuh/wazuh.sh:106]: FIXED

File: src/remoted/manager.c (hunk 3 above)
Comment (DProvinciani) [reply src/remoted/manager.c:94]: FIXED

File: src/remoted/manager.c (hunk 4 above)
Comment (DProvinciani) [reply src/remoted/manager.c:1282]: This is the method used to free the memory allocated for the entries in the `pending_data` hash table. If an entry is removed, we have to release the memory of either the message as well as the entry itself.

File: src/shared/read-agents.c (hunk 5 above)
Comment (DProvinciani) [reply src/shared/read-agents.c:1157]: FIXED

File: src/shared/read-agents.c (hunk 6 above)
Comment (DProvinciani) [reply src/shared/read-agents.c:1161]: FIXED

File: src/shared/read-agents.c (hunk 7 above)
Comment (DProvinciani) [reply src/shared/read-agents.c:1279]: FIXED

File: src/shared/read-agents.c (hunk 8 above)
Comment (DProvinciani) [reply src/shared/read-agents.c:1291]: FIXED

File: src/shared/remoted_op.c (hunk 9 above)
Comment (DProvinciani) [reply src/shared/remoted_op.c:40]: Removed.

File: src/shared/remoted_op.c (hunk 10 above)
Comment (DProvinciani) [reply src/shared/remoted_op.c:57]: All the output pointers were removed.

File: src/shared/remoted_op.c (hunk 11 above)
Comment (DProvinciani) [reply src/shared/remoted_op.c:225]: FIXED

**Review by vikman90 — APPROVED** [review]

**Pull request merged**